### PR TITLE
Upgrade base-sepolia tests for the Fjord upgrade

### PR DESCRIPTION
**Pull request by HrikB**
After the Fjord upgrade went live, the tests validating the estimation algorithm were no longer accurate. This PR updates the calculation for estimation validation. 

### The New Calculation
The SDK will perform an estimate by calling `getL1Fee()` on the gas price oracle. Post-Fjord, this function will use an lz compression algorithm in the EVM emulating the behavior specified by [this](https://github.com/ariya/FastLZ/blob/344eb4025f9ae866ebf7a2ec48850f7113a97a42/fastlz.c#L482-L506) C implementation. Unfortunately, there is no Js counterpart for that specific implementation and the node bindings [available](https://github.com/sparkypants/node-fastlz) are not maintained and out of date.

Instead, for testing, the new calculation will simply hit the on-chain contract through rpc fixed at a Fjord block. This should detect any changes in behavior to the contract.

There are 5 inputs into the `getL1Fee()`:
- The calldata `bytes`
- `baseFee` storage var
- `baseFeeScalar` storage var
- `blobBaseFee` storage var
- `blobBaseFeeScalar` storage var

The test mimics the sdk call by passing the same calldata and using state overrides to manipulate storage vars.

### Random Bytes for Mock Calldata

Since estimation uses a compression algorithm, using a series of repeated bytes would yield an overly-optimistic compression. A series of random bytes were generated for a more pessimistic estimate.

### Bounding vs Equality Check
Currently, the sdk estimation is tested to be within a bound of the fjord-block call. There error bound arises from the fact that the 4 storage vars overriden in the test are being pulled from a different block than the `getL1Fee()` call in the sdk.

image

If the storage vars call in the `beforeEach()` and the sdk call are fixed to the same block, they are equal. While not implemented in this PR, fixing both calls to come from the same block would yield a more potent test.

### Multicall
The 4 storage vars are now multilcalled in the `beforeEach()` to guarantee they are all coming from the same block.

### CODE REVIEW

**Review by yi-sun — APPROVED** [review]

**Comment by ytham** [comment]
Looks like test `client/tests/unit/lib/paymentCalc/84532.test.ts` is failing:
Screenshot 2024-06-06 at 3 29 35 PM

This is unfortunately thrown by Jest now when a test fails (instead of throwing an actual error) and is under investigation by the Jest team. However, you should see the correct error when running this test locally.

**Comment by HrikB** [comment]
> Looks like test `client/tests/unit/lib/paymentCalc/84532.test.ts` is failing: Screenshot 2024-06-06 at 3 29 35 PM
> 
> This is unfortunately thrown by Jest now when a test fails (instead of throwing an actual error) and is under investigation by the Jest team. However, you should see the correct error when running this test locally.

I was getting error locally before but it seems to have gone away. Everything passes locally :/

**Comment by ytham** [comment]
I retried the test and it did indeed pass again, which I guess is probably pointing to some issue that comes up when the L1 gas fee spikes (it looks like there was a spike 1~2 hrs ago)

**Comment by HrikB** [comment]
Do you have a sepolia-base block number to go with it? Perhaps I can repro by fixing the block.

**Comment by ytham** [comment]
Unfortunately I do not; I was only looking at this on https://sepolia.beaconcha.in/gasnow

Screenshot 2024-06-06 at 5 24 38 PM

Do you still have logs from your local test run that failed?

**Comment by HrikB** [comment]
I can derive a block number from this. Let's keep the PR in the air for now. I will monitor the tests to see if I can repro.

File: client/test/unit/lib/paymentCalc/84532.test.ts [hunk 1]
```diff
@@ -71,7 +88,8 @@ describe("PaymentCalc: Base", () => {
     await axiom.prove(inputs);
     const args = axiom.getSendQueryArgs();
 
-    const queryCost = calculateQueryCost(CHAIN_ID, basefee, baseFeeScalar, blobBaseFee, blobBaseFeeScalar, maxFeePerGas, callbackGasLimit, proofVerificationGas);
+    const queryCost =  await calculateQueryCost(publicClient, CHAIN_ID, basefee, baseFeeScalar, blobBaseFee, blobBaseFeeScalar, maxFeePerGas, callbackGasLimit, proofVerificationGas);
+    expect(queryCost).toEqual(BigInt(args?.value ?? 0));
```
Comment (HrikB) [inline client/test/unit/lib/paymentCalc/84532.test.ts:92]: Ok I've isolated the issue. This equality check was left as residue while I was testing for fixed block equality. The equality without fixing blocks is quite unstable.

**Review by yi-sun — APPROVED** [review]

**Review by ytham — APPROVED** [review]
LGTM, thanks for the follow up!

**Pull request merged**